### PR TITLE
Add aarch64 support

### PR DESCRIPTION
**Pull request by b100dian**
This adds the correct lib64 location for the preloaded so.
Also, because a previous version of this might have been installed from b100dian OBS or chum, which used the 64-bit binary in /lib/, that needs to be cleared from ld.so.preload.

<details>
<summary>Old description</summary>
Without bash installed, the busybox shell was erroring out with "Substitution failure".
This changes the script to no longer need `${PIPESTATUS[0]}` which is the cause of the above failure.

Also, when building for aarch64, there is some mismatch between where a lib is packaged and where is looked up. The 'fix' is to currently just hardcode "lib" - that's why I say its' the naive fix, only to unblock building/running.
</details>

### CODE REVIEW

**Review by Olf0 — CHANGES_REQUESTED** [review]
@b100dian, IMO these are bug fixes, but are not related to supporting busybox!
1. `false | true; echo $?` results in 0 (i.e., "O.K") on both, *ash* and *bash*.
2. `set -o pipefail; false | true; echo $?` results in 1 (i.e., "fail") on both, *ash* and *bash*.
* Consequently, your comments "`# for busybox`" (twice: in each of the two files patched) is not correct. 
* Also the addition of ` ; test ${PIPESTATUS[0]} -eq 0` is superfluous, AFAICS.

But you detected at least two places in which `$?` (after a pipeline applying or testing a patch) is never evaluated, and more places in which only one of the cases (failure or success) is checked for!
IMO these shall be rectified, along setting `set -o pipefail` (what you already do).

**Comment by b100dian** [comment]
Thanks @Olf0 for looking into this - the reference to busybox is actually because `${PIPESTATUS[0]}` is erroring out on busybox with "Substitution failed". Does this make sense?

**Review by Olf0 — CHANGES_REQUESTED** [review]
@b100dian, what you labelled as "naive aarch64 support" is truly "naive", but might be an appropriate "quick & dirty hack" to make Patchmanager working on aarch64.
But because 
1. hardcoding library paths, for which RPM macros exist is evil (but I understand that there is no other "quick & dirty" way to let your patched Patchmanager still work on ARMv7-A)
2. this "naive aarch64 support" also comes with a set of fixes, which are completely unrelated (and mislabelled "busybox support")

IMO you shall separate these!

I do not think that this "naive aarch64 support" is appropriate for the "master" branch, in contrast to the fixes labelled "busybox support", when they are enhanced and complete.
Thus these must be separated structurally, in order to be separately applicable.

**Comment by b100dian** [comment]
Agreed - this was just a branch based on the sfos4-fix branch, let me cherry pick the non-aarch64 related commits in a PR against master

**Comment by Olf0** [comment]
> Thanks @Olf0 for looking into this - the reference to busybox is actually because `${PIPESTATUS[0]}` is erroring out on busybox with "Substitution failed". Does this make sense? Let me adda a description to the PR:)

Oh, I did not try it, because it is superfluous on both, *ash* ("busybox") and *bash*, and has not been there before (AFAICS).

But thinking about it, that does make sense, because `${foo[0]}` is a *bash*-only syntax for accessing the first element (i.e., at index 0) of a *bash*-array.  Most other shells do not implement arrays and no othe has this syntax: Hence the error.

P.S.: Mind that all this code shall still run under *bash* (because that is what SailfishOS releases before 3.2.0 use), the task is only to make it busybox's *ash* compatible.

P.P.S.: By reading the error reports WRT Patchmanager 3 on recent SailfishOS releases, I suspect there are more "Bashisms" hidden somewhere in Patchmanager's shell code.
But because I am fighting a [nasty breakage on recent SailfishOS releases of a tool, which I wrote and maintain](https://github.com/Olf0/crypto-sdcard/issues/115), the most I can offer is to review every now and then.  Plus I am only good at shell code!

**Comment by Olf0** [comment]
Sorry, I misunderstood and hence corrected myself, above.

Mind, that you might also want to integrate @CODeRUS existing fixes for SailsfishOS 4.x (although I partially do not like them, because they include a change, which breaks Patchmanager for SaifishOS < 4).

**Comment by CODeRUS** [comment]
> because they include a change, which breaks Patchmanager for SailfishOS < 4).

which one please?

**Comment by Olf0** [comment]
> Agreed - this was just a branch based on the sfos4-fix branch, let me cherry pick the non-aarch64 related commits in a PR against master

Well, thinking about it, this was nitpicking by me: IIRC @CODeRUS clearly stated, that he will not do anything for Patchmanager, any more, hence this is rather about maintaining your own fork in a suitable manner; consequently what is "suitable" is up to you.

**Comment by Olf0** [comment]
> > because they include a change, which breaks Patchmanager for SailfishOS < 4).
> 
> which one please?

Dear @CODeRUS,
a. it is really nice to hear from you.  Also because this is a small indication that you are well.
b. But I am not, as written before, tonight: *Sorry it is late and I have a slight headache*.
c. I will try to look it up again, check and write a comment at MR #48, tomorrow evening.
d. I **must** go to bed, now.

**Comment by Olf0** [comment]
> Agreed - this was just a branch based on the sfos4-fix branch, let me cherry pick the non-aarch64 related commits in a PR against master

Oh, I see!  Another nonsense comment by me, tonight: You already use the `sfos4-fix` branch as a base for your patches.
Really not my day, today. :-(

**Comment by Olf0** [comment]
> > because they include a change, which breaks Patchmanager for SailfishOS < 4).
> 
> which one please?

Answered [here](https://github.com/sailfishos-patches/patchmanager/pull/48#issuecomment-894683952).

**Review by Olf0 — COMMENTED** [review]
This is looking fine.

Side note: BTW, `grep libpreloadpatchmanager /etc/ld.so.preload > /dev/null` is better written as `grep -qF libpreloadpatchmanager /etc/ld.so.preload`, but that is unrelated to this specific change.

P.S:: I wonder why I am able to "approve" this and to "request changes".  Have I become more than a regular Patchmanager user here?  Will check, but for now I will leave this as a simple "comment".

**Comment by b100dian** [comment]
Yeah, but it doesnt work yet https://build.sailfishos.org/package/live_build_log/home:b100dian:patchmanager/patchmanager/sailfish_latest_aarch64/aarch64

**Comment by Olf0** [comment]
Well (just stating the obvious), they really are not there, because
`[  232s] install -m 755 -p libpreloadpatchmanager.so /home/abuild/rpmbuild/BUILDROOT/patchmanager-3.0.65+sfos4.fix.busybox+sfos4.fix.20210913211513.1.g68347dc-1.6.1.jolla.aarch64/usr/lib/libpreloadpatchmanager.so`
`[  232s] : --strip-unneeded /home/abuild/rpmbuild/BUILDROOT/patchmanager-3.0.65+sfos4.fix.busybox+sfos4.fix.20210913211513.1.g68347dc-1.6.1.jolla.aarch64/usr/lib/libpreloadpatchmanager.so`
and 
`[  233s] extracting debug info from /home/abuild/rpmbuild/BUILDROOT/patchmanager-3.0.65+sfos4.fix.busybox+sfos4.fix.20210913211513.1.g68347dc-1.6.1.jolla.aarch64/usr/lib/libpreloadpatchmanager.so`
=> The [install] section uses `/usr/lib/`as library path.
(I am too tired to ponder why; will try to think about it tomorrow.)

Consequently 
`[  234s] error: File not found: /home/abuild/rpmbuild/BUILDROOT/patchmanager-3.0.65+sfos4.fix.busybox+sfos4.fix.20210913211513.1.g68347dc-1.6.1.jolla.aarch64/usr/lib64/libpreloadpatchmanager.so`
during files check (triggered by the `%{_libdir}/libpreload%{name}.so` line in the `%files` section) and again in the subsequent RPMbuild phase.

**Comment by Olf0** [comment]
On first sight, the `%qmake5_install` seems to be the culprit.
I will have to look up, what it really expands to, but this provides a hint: https://github.com/sailfishos/sdk-setup/blob/master/sdk-setup/src/mb2#L4984
Well, you can do that too, by `rpm --showrc | fgrep qmake5_install`
Usually such macros call a shell script.

**Comment by Olf0** [comment]
But the build phase (defined in section `%build`) also puts `libpreloadpatchmanager.so` into `/usr/lib/`, i.e.:
```
%qtc_qmake5 "PROJECT_PACKAGE_VERSION=%{version}"
%qtc_make %{?_smp_mflags}
```

So maybe you might have to adapt the "makefiles" (but there are not any: this is where I lack the understanding for such a "modern" build process).
Or maybe one of the above macros accepts parameters with which one can control the output directory for `libpreloadpatchmanager.so` (which is compiled by these statements, I assume).

**Comment by nephros** [comment]
> But the build phase (defined in section `%build`) also puts `libpreloadpatchmanager.so` into `/usr/lib/`, i.e.:
> 

That's because `/src/preload/preload.pro` tells it to install it there.

Q: Is the library actually supposed to end up in lib64? Some things should go to /usr/lib even in 64bit systems.

**Comment by nephros** [comment]
> Q: Is the library actually supposed to end up in lib64? Some things should go to /usr/lib even in 64bit systems.
> 
> (Note the ld.preload configuration in the .spec also hardcodes /usr/lib not /usr/lib64)

Disregarding whether this is correct, this fixes the compile/install/package error:

https://github.com/nephros/patchmanager/commit/4fa3e0fc06f086b37d3723d0a018fa1320ec25e5

See also the previous commit while we're at it: https://github.com/nephros/patchmanager/commits/sfos4-fix

Builder here:

**Comment by b100dian** [comment]
That was the missing step, thanks @nephros - and now I had to patch the upgrade from lib/ to lib64/ :)

**Comment by b100dian** [comment]
I just tested the upgrade path and it seems to work now. If no-one has anything against, tomorrow I will merge @nephros PRs then branch out `patchmanager3.1` and `merge --squash` the `sfos4-fix` by coderus, then re-target this PR on top of it.

@Olf0 would you want to start a thread on FSO? We talked earlier and agree that "A forum (thread) is IMO the best tool to communicate and organise a Patchmanager developers team in an open and transparent manner" but I'd like to do it on forum.sailfishos.org nonetheless. Would be useful for logging bugs like this one https://forum.sailfishos.org/t/patchmanager-patches-in-koli-4-0-1/4925/143 and the future ones:)

**Comment by Olf0** [comment]
> I just tested the upgrade path and it seems to work now.

Now it really makes sense, per `sed -i "/libpreloadpatchmanager/ d" /etc/ld.so.preload`

But I think this can be simplified and made consistent with the rest of the spec file by using RPM variables:

~~ARCH=$(getconf LONG_BIT)~~
~~SUFFIX=$([[ "$ARCH" == "64" ]] && echo "64" || echo "")~~
sed -i "/libpreload%{name}/ d" /etc/ld.so.preload
echo "%{_libdir}/libpreload%{name}.so" >> /etc/ld.so.preload

**Comment by Olf0** [comment]
Done that (among some other beautifications and clarifications) without altering functionality per https://github.com/b100dian/patchmanager/pull/2

**Comment by Olf0** [comment]
> @Olf0 would you want to start a thread on FSO? We talked earlier and agree that "A forum (thread) is IMO the best tool to communicate and organise a Patchmanager developers team in an open and transparent manner" but I'd like to do it on forum.sailfishos.org nonetheless. Would be useful for logging bugs like this one https://forum.sailfishos.org/t/patchmanager-patches-in-koli-4-0-1/4925/143 and the future ones:)

I think this needs a bit more consideration:
1. I think we need a place for users of Patchmanager to report and discuss their experiences, as the [example you made](https://forum.sailfishos.org/t/patchmanager-patches-in-koli-4-0-1/4925/143)
  - FSO is well suited for this, because most SailfishOS users have an account there.
  - But before starting such a thread, I would like to be able to state, there PM's future development location is (along with its issue tracker etc.).
    Maybe @SfietKonstantin / Basil is ignoring GitHub notifications, I will try to contact him via email this weekend, to check if we can stay here or have to move.
2. "A forum (thread) to communicate and organise a Patchmanager developers team in an open and transparent manner"
   - Thinking about it, it seems to be helpful, if this is not "next door" to the PM users thread.
   - As stated, the infrastructure provided by Jolla regularly undergoes sudden changes (main git repository in its 3rd iteration, bug tracker in its 3rd iteration, build infrastructure in its 2nd iteration with the next iteration announced, but unclear what it may be), which were always handled awkwardly by Jolla.
   - You and @nephros do not seem to like TMO:
      - Can you explain why?
      - Maybe using the wiki of PMs source repo at GH be a better approach:
        - It provides structure.
        - It is where the source code is.

**Comment by nephros** [comment]
I don't have anyrhing against TMO per se, but:

 - I don't have an account there at the moment and am reluctant to add yet another 
 - FSO is where most of modern-day SFOS discussion happens, and even Jolla guys sometimes drop by
 - for both users and deva it's annoying to have a million different places to find information
 - TMO is fullof ancient and confusing old stuff. Great if you're trying to resurrect a N9, but for new stuff not so much


I agree though that the future should be decided in before making an announcement. OTOH PM is broken for many right now, perhaps we should focus on shipping something useful first and then get communication in order.

About the license: obviously it's up to the principle authors to decide, but can you give a reason or two why you'd prefer LGPL?

Code repo: if we can continue here that would be best , if not however, I 'd like to suggest moving away from Microsoft Github - just out of principle. (owner, monopoly, proprietary git extensions  etc).
 -

**Comment by b100dian** [comment]
> it makes sense to open such a user facing thread, when the announcement for a new release (and release location, with a link to it) at Openrepos can be made.

> PM is broken for many right now, perhaps we should focus on shipping something useful first and then get communication in order.

Folks, fully agree. I was excited about the progress:)

**Comment by b100dian** [comment]
Re: TMO
I don't use it, that's why I mostly prefer FSO. Also, forums, as chat services, cannot be quicly evaluated for freedom of expresiveness and even then, things can change quickly. I think code (which we have each a copy of) is more important a.t.m.

**Comment by b100dian** [comment]
> License: obviously it's up to the principle authors to decide, but can you give a reason or two why you'd prefer LGPL?

I'd also be curious. Any free as in freedom license should hold IMO, the project doesn't sound like it's going to be linked against or tivoized.

**Comment by b100dian** [comment]
Code repo: I just asked @CODeRUS to make the `patchmanager31` branch default and he said he'd go for `master`, I made the old master be legacy, and am hoping that `master` is going to be the default branch from now on.
I also updated the existing PRs to point to master when the default is done.

If this is not happening of course I will ask myself I we shouldn't just move away.

**Comment by Olf0** [comment]
@nephros, I think there are a couple of misunderstandings:
- FSO for a PM-*users* thread is fine for me (Vlad already convinced me): I will draft a starting message, when I am done with a draft of the Openrepos page.
- I still think it is preferable not to use Jolla's infrastructure for the PM-related *developer* communication.  I have been burnt by Jolla's permanent changes (especially Gits and TJC → FSO, without the promised migration of content) too many times, and I love to be able to retrace what happened years ago.  That is far less important for bug reports and other user messages, because these tend to be more volatile.
   Side note: Plus for anything technical, except OBS; but we also shall not rely on Jolla's OBS either, because its demise is already announced (but absolutely unclear when): Using the SFOS-SDK locally is supposed to substitute it (LOL).
- The fact that there is a lot of old cruft at TMO is irrelevant, when starting a new thread.

> I agree though that the future should be decided in before making an announcement. OTOH PM is broken for many right now, perhaps we should focus on shipping something useful first and then get communication in order.

I do not comprehend your "OTOH" above: IMO these different aspects can and should be tackled concurrently, and this is what we already do.
While Vlad made PM3 SFOS4-ready (with your help), you nicely addressed some of the old issue reports (thank you!) and I am evaluating the options for infrastructure and preparing an Openrepos page and an initial message for an PM-user thread at FSO.

> License: obviously it's up to the principle authors to decide, but can you give a reason or two why you'd prefer LGPL?

-I will ask Andrey if he is O.K. with an [OSI-approved license](https://opensource.org/licenses), and  expect a "Yes", if he answers at all.
- The LGPL is the "lightest" of the reciprocal licenses.  Do you have a different preference, and if so, why?

> Code repo: if we can continue here that would be best , if not however, I 'd like to suggest moving away from Microsoft Github - just out of principle. (owner, monopoly, proprietary git extensions etc).

But where to?  Gitlab really convinced me to be much worse, and all other solutions are rather niches, plus I wonder how sustainable their hosting is.
While I have been a Microsoft hater for decades (since [Bill Gates' open letter to hobbyists](https://en.wikipedia.org/wiki/Open_Letter_to_Hobbyists)), I have watched their FLOSS strategy developing since 1999 (with their [SFU](https://en.wikipedia.org/wiki/Windows_Services_for_UNIX) experiment), more seriously since 2004 (with [WiX](https://en.wikipedia.org/wiki/WiX)) and fully with the advent of [CodePlex](https://en.wikipedia.org/wiki/CodePlex) in 2006.  But that was always only FLOSS for "the Windows ecosystem".  Ultimately [Satya Nadella](https://en.wikipedia.org/wiki/Satya_Nadella) started to alter Microsoft to a services company using and building upon FLOSS in 2014 and ultimately released [MS Basic as FLOSS in 2020](https://devblogs.microsoft.com/commandline/microsoft-open-sources-gw-basic/): What a kick in Bill Gates' arse, taking the "[open letter to hobbyists](https://en.wikipedia.org/wiki/Open_Letter_to_Hobbyists)"-story into account!
Hence I nowadays trust Microsoft to run Github in a reasonable manner (much more that the Gitlab company, Microsoft tries hard to do nothing wrong), to guarantee sustainability / stability and only a Github→Github migration allows us to transfer more that just the git-tree (issues etc.).  Thus I would rather stay at Github, even if we have to move (which I will evaluate this weekend).

**Comment by Olf0** [comment]
> Code repo: I just asked @CODeRUS to make the `patchmanager31` branch default

That makes sense.

> and he said he'd go for `master`, I made the old master be legacy, and am hoping that `master` is going to be the default branch from now on.

???
This does nor make any sense to me, because I do not comprehend this statement.  What is going to happen now, i.e. what is going to be the default branch?
Can you access this repository's setting (I cannot even look at them).

> I also updated the existing PRs to point to master when the default is done.
> 
> If this is not happening of course I will ask myself I we shouldn't just move away.

IMO, either he agrees to share administrative control or we have to move.

**Comment by b100dian** [comment]
master is now the default branch (copied from patchmanager3).

we can now squash the sfos4 fix by CODeRUS and re-target this

**Comment by b100dian** [comment]
@Olf0 I got full settings now (I didn't, before). It think it was a misunderstanding on chat. First thing I enabled reviews for merging. I like the idea of successor I think you used for the fork, I'll sleep on it

**Comment by Olf0** [comment]
#### Status update
* I sent @CODeRUS an email to his gmail address, covering these topics:
  1. The source code of the https://coderus.openrepos.net/whitesoft/sailversion crawler
  2. [Patchmanager] The code for running the web-service at https://coderus.openrepos.net/pm2 (and sub-pages)
  3. [Patchmanager] A "written" (per email) statement that we can choose an OSI-approved (FLOSS) license, or he chooses such a license now.
  4. [Patchmanager] Adding me in an admin role to the "organisation" Coderus at Transifex.  If he does that, I can provide access rights to others.

* I also played with a new Github account (because I am an engineer and like to evaluate things by trying them out): [Patchmanager](https://github.com/Patchmanager)
  * A Github account owner can name another Github account, which gains control, if the original owner is unreachable (e.g. avoids the internet, is permanently ill, is dead etc.).  This is a recently added function, which I have not noticed before.
    For the Patchmanager account I named Vlad G. (b100dian, who already accepted).
  * I also created source code repositories by "forking" and "importing" extant repositories: This apparently makes no difference, except that "forking" displays the connection to the original repository and easily allows to pose MRs against it; issues, wiki content etc. are not copied when "forking", and "importing" by definition solely clones the source repository.
    So ultimately I deleted all the test repos and [forked the original patchmanager repository](https://github.com/Patchmanager/patchmanager).
  * There is only a single level of access rights an owner of a Github repository can grant others: The "collaborator" status, which allows to perform simple tasks (merging PRs, categorising issues etc.), but provides no access to the repository's settings.  Because the owner status (i.e., full admin) cannot be shared, I created the separate account "Patchmanager" (and did not use my personal one, "Olf0"), so we can work with shared secrets (either a private key or a password) to share access; I still have to evaluate the technical options for this, plus we need a secure channel to transmit such secrets.
    For the patchmanager repo I requested collaborator status from b100dian, nephros and Olf0.
  * Github offers another new function: "Discussions".  These are explicitly aimed at coordinating developers, and seem to be half way between "Issues" and a "Wiki": All texts are written in Markdown, but there is some structure (more than a "flat" issue, but less than a wiki) and tasks can be assigned to accounts.  I actually like the integration with other functionality at Github, and this is where the developers are anyway (no additional login etc.).
  * I also tried to export the [extant SVG icon](https://github.com/sailfishos-patches/patchmanager/blob/master/src/icons/svgs/icon-m-patchmanager2.svg) to PNG for using it as the avatar image for the Patchmanager account and utterly failed: A current Inkscape complains about the ancient Inkscape version with which this icon was supposedly created and loads only fragments of it; this is likely the reason why Github fails to dispaly it, too.  I then considered the [extant icon at Openrepos](https://openrepos.net/sites/default/files/styles/thumbnail/public/packages/12655/icon-patchmanager.png), but that has a very low resolution.  Thus I played 1,5 hours with Inkscape to properly resurrect the foreground elements of the SVG icon (most elements were still there, but without appropriate colors and the layers were collapsed into one, but with objects arranged icorrectly) and used that as Patchmanager's avatar image.  Another task I unwillingly acquired, and the background is completely gone (hence no resurrection, but I have to create it anew).

TL;DR: I evaluated the possibilities for collaboration and tried a migration (per "fork") here at Github.  This seems to work well and is much easier than a migration elsewhere.  I also created a "Patchmanager" account for that, and understood that only a repo owner has full administrative control over it.
If we decide to stay at Github (which is my strong suggestion after this evaluation), we would just have to set a date for switching to the new repository: I can synchronise the new repo with then original one at any time, we simply have to decide, when we stop merging changes into the original one and start using the new one.  For now, I think it is better to finish the ongoing clean-up work here, but to switch before starting to create the 3.1 release.

**Comment by Olf0** [comment]
> master is now the default branch (copied from patchmanager3).

This afternoon the "Patchmanager 3" branch was the default branch: I have seen this is my fork of this repo (some extra settings are copied over when "forking", instead of "importing").

If "master" is default now, what happened to the old "master" branch?

> we can now squash the sfos4 fix by CODeRUS and re-target this

Good, although I do not really understand, why this was not possible before.

> @Olf0 I got full settings now (I didn't, before).

So you can open the page https://github.com/sailfishos-patches/patchmanaget/settings
That is great (but then I would like to understand how Andrey did that, without buying a "team plan" from Github)!

> It think it was a misunderstanding on chat.

So you are in contact with Andrey per Telegram or VK chat?
Good.

> First thing I enabled reviews for merging. I like the idea of successor I think you used for the fork, I'll sleep on it.

For now, it was an experiment, which we may use or discard.
But to have full control over the all the settings, including the ones for the repo owner "Patchmanager" is definitely nicer IMO.

**Comment by Olf0** [comment]
#### Amother status update
I crated an organisation ["PM-developers"](https://github.com/PM-developers), to see, if this can be done free of charge (Github's documentation largely avoids that, but at a few places sounded that it would).  Well, at least for FLOSS, "organisations" are free of charge.  And they do allow to set access rights in may steps.
So I forked the patchmanager repository once more (now for the orinsation "PM-deveolpers") and granted b100dian, nephros and olf full admin access.

**Comment by b100dian** [comment]
> If "master" is default now, what happened to the old "master" branch?

It's called `legacy`

> why this was not possible before.

I didn't have full access and merging (or squashing) to a non-default branch was not my idea of inviting people here to create issues and maybe make PRs

>  I would like to understand how Andrey did that, without buying a "team plan" from Github

I think some functionality is free, as you noted in your forks.

> So you are in contact with Andrey per Telegram or VK chat?

Yes, this is how I initially got the merge access. I tried to keep the messages low as he probably has other stuff to do, and only contacted him for the initial merge request, then for switching the default branch (where there was probably some level of misunderstanding there) and finally if I get the same level of access as his and Sfiet.
Some of these changes (master instead of patchmanager31 and copy the original to legacy) are his suggestions.

This 'fixes' the point 'Do we have owner' for me. I would like to be a bit conservative at first and don't make any other disruptive changes immediately. Tell me what do you think should be changed immediately (like - I required reviews for merging as a first team thing).

**Comment by b100dian** [comment]
As far as I understanding the open points are: Transifex and license

I see you've evaluated GH Discussions a bit @Olf. I'd like to move the developers discussion there instead of PRs:)

On the PR topic: the GH Actions are now enabled by the sfos4-fix included in the default branch and I've included the aarch64 (hopefully, correctly..). If all goes well I think we can enable that check as mandatory for PRs too

**Comment by Olf0** [comment]
Continuing this discussion at https://github.com/sailfishos-patches/patchmanager/discussions/57

**Comment by Olf0** [comment]
> I didn't have full access and merging (or squashing) to a non-default branch was not my idea of inviting people here to create issues and maybe make PRs

O.K., I thought this does not matter, because it is currently just us three looking at this.
Actually I became a bit confused by what went into `sailfishos-patches:master`,  `sailfishos-patches:sfos4-fix` and `b100dian:sfos4-fix`, plus why you changed the default branch multiple times.
BTW, there also is the branch `sailfishos-patches:sfos4`?!?

I will look at the result, when you merged this PR and everything (branches, code) has settled a bit.  I do trust you to do the right things, I but just cannot fully follow: Don't mind (explaining etc.), I will see the outcome.

> This 'fixes' the point 'Do we have owner' for me. I would like to be a bit conservative at first and don't make any other disruptive changes immediately.

Absolutely agreed: The baseline was, "only move when we have to".
Because you seem to have full administrative access for this repo now, the idea of moving might have become obsolete (unless other hurdles appear).

> Tell me what do you think should be changed immediately (like - I required reviews for merging as a first team thing).

As you already switched on discussions, created new build actions, required reviews for merging: These were the essential things.
You could upgrade me to a contributor (or administrator, if you like); currently I cannot perform any administrative work (not even handling issues, except for commenting them).

**Comment by Olf0** [comment]
> BTW, there also is the branch sailfishos-patches:sfos4?!?

Comparing it with `sailfishos-patches:master` or `sailfishos-patches:sfos4-fix` (basically the same result) shows some changes we already have in `b100dian:sfos4-fix` (by nephros), some which make sense to me, and some I do not understand.
https://github.com/sailfishos-patches/patchmanager/compare/master...sfos4

**Comment by b100dian** [comment]
`sfos4` is reminiscent of closed https://github.com/sailfishos-patches/patchmanager/pull/47 - I deleted it, can be restored the PR interface

**Comment by b100dian** [comment]
The default branch matters precisely because if you have the `.github/workflows` in it it will use GH actions

sfos4-fix can be also be deleted, as it can be restored from https://github.com/sailfishos-patches/patchmanager/pull/48
I'll leave it for now because it helps with the explanation:

Changes I've done:
- copied master to legacy: `git push origin master:legacy`
- copied patchamanager3 to master
- Squashed (not merged) the sfos4 fix. You can see there is only one commit in master for that: f9d0ea51642fe9abca3a3e4be4a40744efcf5903
- Then changed the target of this PR to be master. Realized it is not merging because it was based on sfos4-fix, but sfos4-fix was just squashed into master (different history:).
- locally `git rebase origin/sfos4-fix --onto master` got my commits (and yours) replayed on the master (the command is approximate, the rebase did not include the 'busybox support' which was already in master..)
- manually inspected `git diff HEAD..master` to match this PR (this is when I switched it back to point to sfos4-fix)
- was satisfied, pushed the changes (--force), then switched the PR back to target master.

Then I added aarch64 to the GH workflow.

This PR still needs review you know:)

I've upgrade you contributor, I was also under the impression that you are part of some team for reviews worked

**Comment by Olf0** [comment]
> `sfos4` is reminiscent of closed #47 - I deleted it, can be restored the PR interface

O.K. (actually I hate to see stuff deleted, without knowing for sure, that it is obsolete or superfluous).
As said, it [already contained the fixes](https://github.com/sailfishos-patches/patchmanager/pull/47/commits/aaa59d92bfda960e41abc82572b79b2f534c85a3#diff-aec10608f10b9992c14a18ee9b82a7c53c6476d40eccf35cc152889d51770173L58) @nephros has rediscovered.

I just kindly ask you and / or @nephros to check, if its two commits contain other relevant or useful changes (primarily the first one): 
https://github.com/sailfishos-patches/patchmanager/pull/47/commits/aaa59d92bfda960e41abc82572b79b2f534c85a3
https://github.com/sailfishos-patches/patchmanager/pull/47/commits/75217c3cac32c5c3faf644d08c865663c53bf951

P.S.: Will start reviewing now.

**Review by Olf0 — APPROVED** [review]
Looking good.

**Review by Olf0 — APPROVED** [review]
Approve a second time?  O.K.

**Comment by b100dian** [comment]
Did you accept the invitation for contributors?;)

**Comment by Olf0** [comment]
> Did you accept the invitation for contributors?;)

There was none!  Just checked again in my GH-notifications.
But I am labelled as "Contributor" (just like nephros, and you as "Contributor" and "Author").

The documentation is not really helpful, either: https://docs.github.com/en/github/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/about-pull-request-reviews and liked pages there.
It details the roles at "organisations" (where I have seen the more granular rights), but I should have "Write access" now.   Will try to create a new branch for testing.

**Comment by Olf0** [comment]
According to the documentation "collaborators" do have "write access" at "personal repositories": https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-github-user-account/managing-user-account-settings/permission-levels-for-a-user-account-repository#collaborator-access-for-a-repository-owned-by-a-user-account

But as a matter of fact, my approval turned the state of this PR to "red"!

**Comment by b100dian** [comment]
Hmm 
image

**Comment by Olf0** [comment]
BTW, it seems that PRs are "automatically merged" after a successful review, i.e. that a reviewer's O.K. triggers merging of a PR.
Such automatisms give me a bad feeling, I would rather have a human to click on "merge now".  E.g., meanwhile something may have changed in an (technically) non-conflicting way in the branch to merge to, which renders parts of the merge senseless.

**Comment by Olf0** [comment]
I will log out and then back in: Maybe the notification then appears.

**Comment by b100dian** [comment]
Sounds familiar but doesn't seem enabled, where did you see this?
image

**Comment by Olf0** [comment]
1. Still no notification for the contributor status.
   But notifications for your new messages do appear.
2. Good that it is unchecked.
I have seen this here, in the third box under the comments:
> Merging is blocked
Merging can be performed automatically with 1 approving review.

**Comment by b100dian** [comment]
I will approve the remose timer PR to see if it has this outcome:)

**Comment by Olf0** [comment]
> And maybe the email in your spam folder.

From my experiments with roles and orgs, I know that all notification show up in the web-frontend, hence I looked only there: This one did not.
=> Accepted & Retrying

**Comment by Olf0** [comment]
O.K., I could merge now.
I will leave it to you, if you want a single squashed commit or a regular merge of all commits.
Or do you want me to try merging?

**Comment by b100dian** [comment]
Yeah, squash it

On Tuesday, September 21, 2021, olf wrote:
> O.K., I could merge now.
> I will leave it to you, if you want a single squashed commit or a regular merge of all commits.
> Or do you want me to try merging?
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/sailfishos-patches/patchmanager/pull/49#issuecomment-92331705

-- 
Sent from my Sailfish device

**Comment by Olf0** [comment]
> Yeah, squash it

Done.

P.S.: In the personal GH-settings for notifications, in the section "Actions" there is the section "Notifications for workflow runs on repositories set up with GitHub Actions." with three choices: "Email" was set, but "Web" was not.
Maybe setting "Web" (as done now) would have made the difference.

**Comment by Olf0** [comment]
... and my fears of auto-merging were just based on bad wording in GH's web-frontend.  And you already have schown that this setting is off.

**Comment by Olf0** [comment]
> I just kindly ask you and / or @nephros to check, if its two commits contain other relevant or useful changes (primarily the first one):
https://github.com/sailfishos-patches/patchmanager/commit/aaa59d92bfda960e41abc82572b79b2f534c85a3
https://github.com/sailfishos-patches/patchmanager/commit/75217c3cac32c5c3faf644d08c865663c53bf951

O.K., looking at it more closely, I believe that all relevant changes in the first commit were carried over by Coderus from the sfos4 to the sfos4-fix branch (with some small enhancements).
Hence I also conclude that sfos4 was just a stale branch as part of PR #47, as Vlad already pointed out.

And the sfos4-fix branch should be deleted now, because what I merged into master tonight was based on it (and it sure did not receive any further commits).
But as I want to keep the "4 eyes" principle, this is my suggestion and somebody else can do this or I will do it, if someone else states "O.K.".

**Comment by b100dian** [comment]
image

> sfos4-fix can be also be deleted, as it can be restored from #48

@Olf0 sleep, tomorrow we patch from /var/run :stuck_out_tongue_winking_eye:

**Comment by Olf0** [comment]
Oh, took me a while to have an idea [what this may address](https://github.com/sailfishos-patches/patchmanager/issues/55#issuecomment-923901147). :thinking:

It is nice to see you and @nephros so motivated and active, but I am usually quite tired, when taking a look at this after work. :sleepy:
P.S.: Currently, this is stealing much of my spare time.

**Comment by b100dian** [comment]
We are all working.. -  @nephros seems to have a head start now, but we may jump in and fill in the gaps when he's underwater:)
